File: design-pattern-jvm/app/src/test/java/design/pattern/java/patterns/behavioral/state/StateTest.java
package design.pattern.java.patterns.behavioral.state;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

class StateTest {
  @Test
  void ShouldChangeToGreenWhenRedChange() {
    TrafficLight trafficLight = new TrafficLight(new RedLight());
    TrafficLightState lightObtained = trafficLight.changeLight()
      .getCurrentState();
    Assertions.assertTrue(lightObtained instanceof GreenLight);
  }
  @Test
  void ShouldChangeToRedWhenYellowChange() {
    TrafficLight trafficLight = new TrafficLight(new YellowLight());
    TrafficLightState lightObtained = trafficLight.changeLight()
      .getCurrentState();
    Assertions.assertTrue(lightObtained instanceof RedLight);
  }
  @Test
  void ShouldChangeToYellowWhenGreenChange() {
    TrafficLight trafficLight = new TrafficLight(new GreenLight());
    TrafficLightState lightObtained = trafficLight.changeLight()
      .getCurrentState();
    Assertions.assertTrue(lightObtained instanceof YellowLight);
  }
}

File: design-pattern-jvm/app/src/test/kotlin/design/pattern/kotlin/patterns/behavioral/state/StateTest.kt
package design.pattern.kotlin.patterns.behavioral.state

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class StateTest {
  @Test
  fun ShouldChangeToGreenWhenRedChange() {
    val trafficLight = TrafficLight(RedLight())
    val lightObtained = trafficLight.changeLight().currentState()
    Assertions.assertTrue(lightObtained is GreenLight)
  }
  @Test
  fun ShouldChangeToRedWhenYellowChange() {
    val trafficLight = TrafficLight(YellowLight())
    val lightObtained = trafficLight.changeLight().currentState()
    Assertions.assertTrue(lightObtained is RedLight)
  }
  @Test
  fun ShouldChangeToYellowWhenGreenChange() {
    val trafficLight = TrafficLight(GreenLight())
    val lightObtained = trafficLight.changeLight().currentState()
    Assertions.assertTrue(lightObtained is YellowLight)
  }
}
